src/multipart.rs

```diff
@@ -0,0 +1,123 @@
+use super::body::{Body, BodyKind};
+use super::{Error, ErrorKind, Result};
+use mime::Mime;
+use multipart::client as mp;
+use std::fmt;
+use std::io::{copy, prelude::*, Cursor, Error as IoError, Result as IoResult};
+
+/// A file to be uploaded as part of a multipart form.
+#[derive(Debug, Clone)]
+pub struct MultipartFile {
+    name: String,
+    file: Vec<u8>,
+    filename: Option<String>,
+    mime: Option<Mime>,
+}
+
+impl MultipartFile {
+    /// Constructs a new `MultipartFile` from the name and contents.
+    pub fn new(name: impl AsRef<str>, file: impl AsRef<[u8]>) -> Self {
+        let name = name.as_ref().to_owned();
+        let file = file.as_ref().to_owned();
+        Self {
+            name,
+            file,
+            filename: None,
+            mime: None,
+        }
+    }
+
+    /// Sets the MIME type of the file.
+    ///
+    /// # Errors
+    /// Returns an error if the MIME type is invalid.
+    pub fn with_type(self, mime_type: impl AsRef<str>) -> Result<Self> {
+        let mime_str = mime_type.as_ref();
+        let mime: Mime = match mime_str.parse() {
+            Ok(mime) => mime,
+            Err(()) => return Err(Error(Box::new(ErrorKind::InvalidMimeType(mime_str.to_string())))),
+        };
+        Ok(Self {
+            mime: Some(mime),
+            ..self
+        })
+    }
+
+    /// Sets the filename of the file.
+    pub fn with_filename(self, filename: impl AsRef<str>) -> Self {
+        Self {
+            filename: Some(filename.as_ref().to_owned()),
+            ..self
+        }
+    }
+}
+
+/// A builder for creating a `Multipart` body.
+#[derive(Debug, Clone, Default)]
+pub struct MultipartBuilder {
+    text: Vec<(String, String)>,
```

Comment:
I wonder if it would be possible to avoid allocation here, at least until we build the `Multipart` struct.

```rust
pub struct MultipartFile<'key, 'data> {
    name: &'key str,
    file: &'data [u8],
    filename: Option<&'key str>,
    mime: Option<Mime>,
}

pub struct MultipartBuilder<'key, 'data> {
    text: Vec<(&'key str, &'data str)>,
    files: Vec<MultipartFile<'key, 'data>>,
}